databuilder/query_model.py

```diff
@@ -365,38 +359,83 @@ class DomainMismatchError(Exception):
     ...
 
 
-PATIENT_DOMAIN = object()
+def validate_input_domains(node):
+    # The domain of a Frame or Series can be thought of as the set of its primary keys.
+    # This determines which other Frames or Series it can be validly composed with. The
+    # rules we want to enforce are:
+    #
+    #  * One-row-per-patient data can be composed with anything: by design every table
+    #    has a patient_id column and so data defined purely in terms of patient_id
+    #    can be joined with anything.
+    #
+    #  * Many-rows-per-patient data drawn from different tables can never be combined
+    #    — at least, not until we add an explicit JOIN operation.
+    #
+    #  * Many-rows-per-patient data drawn from a single underlying table can be combined
+    #    as long as any filters applied to the data can be arranged in a linear
+    #    sequence. That is, it's legal to combine `Table A -> Filter B -> Filter C` and
+    #    `Table A -> Filter B`. But it's not legal to combine `Table A -> Filter D` and
+    #    `Table A -> Filter E`. There's never a need to use such constructions and they
+    #    will almost always be the result of user error. It's better to reject them
+    #    immediately than give the user unexpected results at query time.
+    #
+    # We enforce this by modelling domains as a hierarchy. At the root is the patient
+    # domain, each table has an unique domain descending from this, and each filter
+    # operation creates a new domain descending from the domain of its source. A valid
+    # operation is one where the domains of its inputs are "nested" in the sense that
+    # each domain is an ancestor or descendant of every other.
+    #
+    # We use sets to implement this hierarchy: we start with the patient domain as a set
+    # containing a single unique value, and we create descendent domains by forming the
+    # union of the parent domain with a new unqiue value. The `issubset()` relation then
+    # gives us the "is domain ancestor" semantics we want.
```

Comment:
This is neat, but would using a list would be clearer?  Instead of `issubset()`, you'd have to write an `isprefix()` function.

Reply:
I did experiment with this, but it's not just an `isprefix()` function you need to write: you also need to get the sorting working correctly, which means writing a comparison function and then passing that through `cmp_to_key()` to make it a key function you can give to `sorted()`. Also, the comparison function you end up writing has to lie and pretend that all domains are comparable so you need a comment explaining why that's OK in this context. Whereas sets give us exactly the semantics we want out of the box. So I feel like overall it's simpler, though the trick does require some explanation.